src/mem/Allocator.h

```diff
@@ -29,38 +36,87 @@ class Allocator : public Interface
      */
     virtual size_t getStats(PoolStats &) = 0;
 
-    virtual PoolMeter const &getMeter() const = 0;
-
     /// provide (and reserve) memory suitable for storing one object
-    virtual void *alloc() = 0;
+    void *alloc() {
+        if (++countAlloc == FlushLimit)
+            flushCounters();
+        return allocate();
+    }
 
     /// return memory reserved by alloc()
-    virtual void freeOne(void *) = 0;
-
-    /// brief description of objects returned by alloc()
-    virtual char const *objectType() const { return label; }
-
-    /// the size (in bytes) of objects managed by this allocator
-    virtual size_t objectSize() const = 0;
+    void freeOne(void *obj) {
+        assert(obj != nullptr);
+        (void) VALGRIND_CHECK_MEM_IS_ADDRESSABLE(obj, objectSize);
+        deallocate(obj);
+        ++countFreeOne;
+    }
 
     /// the difference between the number of alloc() and freeOne() calls
-    virtual int getInUseCount() = 0;
+    virtual int getInUseCount() const { return meter.inuse.currentLevel(); }
 
     /// \see doZero
     void zeroBlocks(const bool doIt) { doZero = doIt; }
 
-    int inUseCount() { return getInUseCount(); } // XXX: drop redundant?
-
     /// XXX: Misplaced -- not all allocators have a notion of a "chunk". See MemPoolChunked.
     virtual void setChunkSize(size_t) {}
 
+    virtual bool idleTrigger(int shift) const = 0;
+
+    virtual void clean(time_t maxage) = 0;
+
+    /**
+     * Flush temporary counter values into the statistics held in 'meter'.
+     * Performs some CPU intensive calculations which have been deferred for performance.
+     */
+    void flushCounters() {
+        if (countFreeOne) {
+            meter.gb_freed.update(countFreeOne, objectSize);
+            countFreeOne = 0;
+        }
+        if (countAlloc) {
+            meter.gb_allocated.update(countAlloc, objectSize);
+            countAlloc = 0;
+        }
+        if (countSavedAllocs) {
+            meter.gb_saved.update(countSavedAllocs, objectSize);
+            countSavedAllocs = 0;
+        }
+    }
+
     /**
      * \param minSize Minimum size needed to be allocated.
      * \retval n Smallest size divisible by sizeof(void*)
      */
     static size_t RoundedSize(const size_t minSize) { return ((minSize + sizeof(void*) - 1) / sizeof(void*)) * sizeof(void*); }
 
+public:
+
+    /// the number of calls to Mem::Allocator::alloc() since last flush
+    size_t countAlloc = 0;
+
+    /// the number of malloc()/calloc() calls avoided since last flush
+    size_t countSavedAllocs = 0;
+
+    /// the number of calls to Mem::Allocator::freeOne() since last flush
+    size_t countFreeOne = 0;
+
+    // XXX: no counter for the number of free() calls avoided
+
+    /// brief description of objects returned by alloc()
+    const char *label;
```

Comment:
Please make this data member const if possible. It clarifies the code, and it is much easier to justify keeping it public this way:

```suggestion
    const char *const label;
```